Add a Physics for THM compressible Euler flow model

GiudGiud opened the pull request:
I hope to figure out at least with VACE:
- [x] channels
- [x] inlets/outlets
- [x] simple junctions

with FV
- [x] channels
- [x] inlets/outlets
- [x] simple junctions
- [x] orienting pipes
- [ ] heat
- [ ] scalars

I ll remove the WCNSFV part from the PR and put it into another PR. I m just trying things here to make sure the design will suit our needs

GiudGiud committed:
Add a base class for THM flow model physics

GiudGiud committed:
Add derived class for VACE, THM single phase flow model

GiudGiud committed:
Vectorize base thm physics class on flow channels

GiudGiud committed:
Have an option to retrieve flow channels from parameters

GiudGiud committed:
Add ability to add inlets to base THM physics

GiudGiud committed:
Add mdot T inlet to VACE physics

GiudGiud committed:
Add base class for flow boundaries usuing physics

GiudGiud committed:
Add a class for flow channels using Physics

GiudGiud committed:
Add a class for mdot - T inlet using Physics

GiudGiud committed:
Adapt base flow components classes for missing flow channel

GiudGiud committed:
Add outlets to THM flow physics

GiudGiud committed:
Add a fixed pressure physics outlet

GiudGiud committed:
Generalize 1 phase closure to accept other types of channels

GiudGiud committed:
Add a new task to avoid THM initialization task-ordering issue

GiudGiud committed:
Silence an info message on THM channel setup

GiudGiud committed:
Add an example

GiudGiud committed:
only add gravity term for non zero gravity

GiudGiud committed:
Add first stab at a WCNSFV physics

GiudGiud committed:
Turn example into a test for VACE

joshuahansel commented on modules/thermal_hydraulics/include/physics/THMWCNSFVFlowPhysics.h:
```diff
@@ -0,0 +1,69 @@
+/**
+ * Sets up the single-phase flow equations using Euler's equations and a RDG discretization
+ */
+class THMWCNSFVFlowPhysics : public virtual ThermalHydraulicsFlowPhysics, public virtual WCNSFVFlowPhysics
```
This will need a test of course

GiudGiud replied:
oh this is like super not ready. will do

joshuahansel replied:
Yeah I realized that later in the review 😄

joshuahansel commented on modules/thermal_hydraulics/src/physics/THMVACESinglePhaseFlowPhysics.C:
```diff
@@ -0,0 +1,662 @@
+
+    // boundary flux user object
+    // we add them in addBCs for convenience
+    if (boundary_type == InletTypeEnum::MdotTemperature)
```
Would someone be able to add a new BC formulation in their application without modifying THM?

GiudGiud replied:
yes, by using a XYZFlowChannel instead of the Physics one
or by creating a new physics

joshuahansel commented on modules/thermal_hydraulics/src/physics/THMVACESinglePhaseFlowPhysics.C:
```diff
@@ -0,0 +1,662 @@
+}
+
+void
+THMVACESinglePhaseFlowPhysics::addInletBoundaries()
```
Do we need to distinguish inlet/outlet, or can we just have a generic `addFlowBCs()` or whatever?

GiudGiud replied:
modeling after WCNSFV here. in wcnsfv we do need to distinguish

joshuahansel reviewed (changes requested):
Just some initial thoughts to get started.

GiudGiud commented on modules/thermal_hydraulics/test/tests/physics/steady.i:
```diff
@@ -0,0 +1,101 @@
+  []
+
+  [pipe]
+    type = PhysicsFlowChannel
```
yes, I started differently but I ll correct course

GiudGiud commented on modules/thermal_hydraulics/test/tests/physics/steady.i:
```diff
@@ -0,0 +1,101 @@
+  [inlet]
+    type = PhysicsInletMassFlowRateTemperature
+    input = 'pipe:in'
+    physics = 'all'
```
good point

GiudGiud commented:
thanks for the early feedback!

GiudGiud committed:
Switch to virtual inheritance for physics to allow a physics to derive both from THM & NS physics

GiudGiud committed:
Set dimension to 1 for TH physics
Suppress boundary parameters

GiudGiud committed:
Add support for setting BCs from components, with a default set at the Physics

GiudGiud committed:
Remove trivial override of addUserObject
This is better, if someone adds addUO in the TH physics base class, it wont compile until they clear the non-trivial override

GiudGiud committed:
Forward adding variable from problem to simulation

GiudGiud committed:
Add an option for adding inlet & outlet BCs to wcnsfv physics from the components

GiudGiud committed:
Add option to specify lower dimension for RC UO and flux BCs

GiudGiud committed:
Centralize list of valid WCNSFV inlet/outlet types

GiudGiud committed:
Make inherited object creation routines protected
Add more options for adding inlet boundaries
Use _has_flow_boundaries in RC UO checks

GiudGiud committed:
Add variables used in THM problems consistently through the Simulation class

GiudGiud committed:
Move addFEBCs down to the specialized VACE class as the WCNSFV THM physics does not use it

GiudGiud committed:
Silence inlet area PPs output by default

GiudGiud committed:
Use the flow channels to specify the Physics, rather than a flow_channel param in the physics
Use the flow channel physics for their connected boundaries
Use a new flow model ID when using Physics
Register 1 phase closures to new flow model id since they should always work

GiudGiud committed:
Adapt VACE and wcnsfv single-channel input for new channel-driven physics input

GiudGiud committed:
Add support for a TH free boundary using Physics

GiudGiud committed:
Add additional APIs to junctions to be able to add them from Physics
Add a base class for Physics-using junctions
Add the one-to-one derived junction

GiudGiud committed:
Add support for adding Junction to THM physics
Add 1-to-1 to VACE

GiudGiud committed:
Add option to specify lower dimension for outflow BCs

GiudGiud committed:
Add functors and functormaterials to take fixed functor values

GiudGiud committed:
Add 1-to-1 flow junction between 1D channels

GiudGiud committed:
Move adding physics into init() routine to perform it early enough

GiudGiud committed:
Better param name for an internal routine

GiudGiud committed:
Add junction test for VACE

GiudGiud committed:
Add junction test for WCNSFV

GiudGiud committed:
Add support for setting gravity and friction terms from outside the physics

GiudGiud committed:
Move friction and gravity info from components to physics data

GiudGiud committed:
Add PPing and try shifting channel on its side

GiudGiud committed:
Modify the mesh to get the right area for flow channels

GiudGiud committed:
Enable coupling auxvariables into AD functor materials as non-AD functors

GiudGiud committed:
Add APIs for modifying eleminfo and faceinfo underlying data

GiudGiud committed:
Create flow channel orientation as a variable

GiudGiud committed:
WIP: typo fixes + try a direction variable

GiudGiud committed:
WIP: move dimension setting into derived class

GiudGiud committed:
WIP Add a functor material for the dimension
Hacky: modify the fv info data

GiudGiud committed:
debugging wcnsfv oriented channel

GiudGiud committed:
Add APIs to hack face info and elem info

GiudGiud committed:
Dont add diffusion kernels if viscosity is 0

GiudGiud committed:
Align all faceinfo information with X axis

GiudGiud committed:
Add capability to add a boundary to WCNSFV fluid heat transfer physics

GiudGiud committed:
Add a Physics to solve the heat equations using WCNSFV

GiudGiud committed:
Use conservative flux BCs for temperature

GiudGiud committed:
Add some partial functor material support to closures

GiudGiud committed:
Add all functor material for THM closure (f and htc)

GiudGiud committed:
Add switches to create functor mateirals in THM closure

GiudGiud committed:
Add an option to add wall heat transfer boundaries programmatically for WCNSFV

GiudGiud committed:
Add ability for adding external heat sources to NSFV fluid heat transfer physics

GiudGiud committed:
Add a conversion functor for treating surface terms with a volume kernel

GiudGiud committed:
Add a functor coef for FVCoupledForce for applying scaled heat fluxes

GiudGiud committed:
Add support for heat transfer components in THM-wcnsfv and base physics

GiudGiud committed:
Add components for applying a heat transfer on channels using Physics

GiudGiud committed:
Add support for heat transfer components in VACE

GiudGiud committed:
Add material property output to closures

GiudGiud committed:
Let closure work with both 1phase and physics-heat transfer

GiudGiud committed:
Add a test for using a heated channel using physics

GiudGiud committed:
Add a test for using a heated channel using VACE physics

GiudGiud committed:
Add ICs for scalars on flow channels
Rename channel test inputs for convenience

We could make this more general, as in ICs for any variable

GiudGiud committed:
Add 'hasBlocks' to Physics. This is yet another API from BlockRestrictable
Maybe it s time we make BlockRestrictableDynamic

GiudGiud committed:
Add a general flow boundary to be able to set any variable value / equation fluxes

GiudGiud committed:
Add support for the general flow boundary in flow and energy WC physics

GiudGiud committed:
Add a Physics for Thermal hydraulics with WCNSFV Scalar transport

GiudGiud committed:
Rework NSFV scalar transport physics to use a map for the inlet boundaries

GiudGiud committed:
Add an input for single channel scalar advection (inlet, channel, outlet)

GiudGiud committed:
Add base support for scalar transfers in flow channel

GiudGiud committed:
Add base support for scalar transfers in closures

GiudGiud committed:
Add base classes for scalar transfer

GiudGiud committed:
Add a functor material to forward values from UOs through functors

GiudGiud committed:
Add a component/transfer connector for scalars from user-specified functors

GiudGiud committed:
Add a coefficient for volume to area conversions to easily add h(a1 - a2) type terms
as volumetric sources in 1D flow channels

GiudGiud committed:
Do not error on missing implementation for missing boundary term
It's likely we ought to have a "surface" physics to handle that rather than
do everything in the TH physics

GiudGiud committed:
Add kernesl for volumetric contribution of scalar surface transfer terms

GiudGiud committed:
Add the ability to add external sources to a NS scalar transport physics

GiudGiud committed:
Add the blocks and set the dimension in the THM Physics base class

GiudGiud committed:
Add example with a scalar source

GiudGiud committed:
Add example with a scalar diffusing in a violume

GiudGiud committed:
Add docstrings for nodal connector functor materials

GiudGiud committed:
Implement Lise and Josh's design advice:
- use fluid properties from the flow channels
- use simple names, no acronyms, for the physics syntax

GiudGiud committed:
Misc test fixups + missing dependency

GiudGiud committed:
Adapt linear fv scalar transport physics to new containers for inlet functors and sources

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2618088/), step Clang format on b79133b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28528/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28528/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2974c54c5269c9b763547954236d33e3a55a8710`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2618088/), step Fixup commit check on b79133b wanted to post the following:

The following commit(s) contain fixups:

- b79133b5b8: fixup! Adapt linear fv scalar transport physics to new containers for inlet functors and sources
- eee77075b6: fixup! Add the ability to add external sources to a NS scalar transport physics
- cf221597f3: fixup! Rework NSFV scalar transport physics to use a map for the inlet boundaries
- 7abcff1cf4: fixup! Switch to virtual inheritance for physics to allow a physics to derive both from THM & NS physics
- 90e1d84a62: fixup! Make inherited object creation routines protected Add more options for adding inlet boundaries Use _has_flow_boundaries in RC UO checks
- c5b55dad13: fixup! Remove NSFVAction : covered by WCNSFV Physics Remove FlowComponentNS: covered by FileMeshPhysicsComponent refs #25642
- d28d5a5b9f: fixup! Add support for setting BCs from components, with a default set at the Physics

GiudGiud committed:
Switch a vector of MooseEnum for passive scalar inlets
Add checks on the sizes of the vectors for passive scalar sources
Clean up logic and only search the boundary type once

GiudGiud committed:
Rework parameter checks again. Even though we check
the boundary types/functors before creating the objects, we need to still check user parameters

GiudGiud committed:
Adapt to new API names in Physics

GiudGiud committed:
Move routines from protected to private to be able to use them in NSFV THM

GiudGiud committed:
Add a test with two channel and two fluid properties for VACE

GiudGiud committed:
Delay variable-system integrity checks because adding variables is
delayed for THM problems

GiudGiud committed:
Update the friction coefficient creation to make it easier
to match closure and kernels

GiudGiud committed:
Rename all the components to remove the 'Physics' prefix per
recommendations

GiudGiud committed:
Remove old deprecated components

GiudGiud committed:
Update tests without the deprecated components

GiudGiud committed:
Make sure variables added in the input files with no components still get added

GiudGiud committed:
Add a test with two channels and two FPs for WCNSFV

GiudGiud committed:
Add convenience option to show the actual existing functors

GiudGiud committed:
Add a component for a heat transfer from a heat flux functor

GiudGiud committed:
Add an input demonstrating a loop with VACE Physics

GiudGiud committed:
Add an input demonstrating a natural circulation loop with WCNSFV physics

GiudGiud committed:
Add a test for a pipe in every direction

GiudGiud committed:
Dont rely on component_names and flow_channel_names being aligned
Add a check on the closures to make sure we are creating the right type of materials

GiudGiud committed:
Add test for closure material type check

GiudGiud committed:
Add an API for knowing the name of the functor providing the wall heat flux

GiudGiud committed:
Remove deprecated components as they have good names which could be used by a general physics-based component

GiudGiud committed:
Add a volume junction for physics

GiudGiud committed:
Implement the volume junction inside the vace-1p thm physics

GiudGiud committed:
Add volume junction pump for physics: currently only pump

GiudGiud committed:
Allow physics to create the vace 1phase pump

GiudGiud committed:
Add inlet for v-T and stagnation p-T to help:
-coupling to SAM
-modeling pressurizer

GiudGiud committed:
Add support in VACE physics for stagnation pT and vel-T inlets

GiudGiud committed:
Add a solid wall to Physics-based THM syntax

GiudGiud committed:
Add a junction for parallel channels for the THM physics syntax

GiudGiud committed:
Add the parallel flow channels junction in the VACE Physics

GiudGiud committed:
Add test inputs for junctions using Physics
pump / volume / parallel / one to one

GiudGiud committed:
Add an option for variable cp for TP function-fp
Use a different definition to compute cp from cv
refs #21245

GiudGiud committed:
Add unit tests for the new variable cp option
refs #21245

GiudGiud committed:
Add the heated perimeter as a multiplying factor on the local heat source

GiudGiud committed:
Add a case showing we meet the analytical solution for non-constant cp in 1D

GiudGiud committed:
Minor fixes for VACE for pump loop
Enable THM closures for physics flowchannel

GiudGiud committed:
Add support for vel-T inlet in THM WCNSFV physics

GiudGiud committed:
Add a loop that mimicks the tutorial forced loop
Re-organize loops together

GiudGiud committed:
Clean up WCNSFV physics:
- make weakly compressible the default
- give up on the forwarding functors idea
